File: src/lib/importers/github.ts
/**
 * GitHub importer (REST v3). Talks to `<host>/repos/<owner>/<repo>` and
 * `<host>/repos/<owner>/<repo>/releases`; `host` defaults to https://api.github.com and is
 * the `/api/v3` base on GitHub Enterprise.
 */
import type { GithubSourceConfig } from '../config/schema';
import type { Release, ReleaseAsset } from '../data/schema';
import {
  JsonClient,
  absoluteUrl,
  byteSize,
  defaultUserAgent,
  nullIfEmpty,
  sortAssets,
  sortReleases,
  stringArray,
  toIsoDate,
  toSpdx,
} from './http';
import type { ImportedReleases, Importer, ImportedRepoMeta, ImporterContext } from './types';

/** Releases per request; GitHub's maximum, so most repos need a single round trip. */
const PER_PAGE = 100;

/** The subset of `GET /repos/{owner}/{repo}` this importer reads. */
interface GithubRepo {
  name?: string | null;
  description?: string | null;
  homepage?: string | null;
  html_url?: string | null;
  clone_url?: string | null;
  default_branch?: string | null;
  topics?: unknown;
  license?: { spdx_id?: string | null } | null;
  has_issues?: boolean;
  is_template?: boolean;
  archived?: boolean;
}

/** The subset of `GET /repos/{owner}/{repo}/releases` this importer reads. */
interface GithubRelease {
  tag_name?: string | null;
  name?: string | null;
  body?: string | null;
  html_url?: string | null;
  draft?: boolean;
  prerelease?: boolean;
  published_at?: string | null;
  created_at?: string | null;
  author?: { login?: string | null } | null;
  assets?: GithubAsset[] | null;
}

interface GithubAsset {
  name?: string | null;
  browser_download_url?: string | null;
  url?: string | null;
  size?: number | null;
  content_type?: string | null;
}

export class GithubImporter implements Importer {
  readonly provider = 'github' as const;

  protected readonly source: GithubSourceConfig;
  protected readonly fetchImpl: typeof fetch;
  protected readonly token: string | null;
  protected readonly userAgent: string;
  private readonly client: JsonClient;

  constructor(source: GithubSourceConfig, ctx: ImporterContext = {}) {
    this.source = source;
    this.fetchImpl = ctx.fetchImpl ?? globalThis.fetch;
    this.token = ctx.token ?? null;
    this.userAgent = ctx.userAgent ?? defaultUserAgent();
    this.client = new JsonClient({
      auth: 'bearer',
      fetchImpl: this.fetchImpl,
      token: this.token,
      userAgent: this.userAgent,
      accept: 'application/vnd.github+json',
    });
  }

  async fetchMeta(): Promise<ImportedRepoMeta> {
    const repo = await this.client.get<GithubRepo>(this.repoUrl());
    const webUrl = nullIfEmpty(repo.html_url) ?? this.fallbackWebUrl();
    return {
      name: nullIfEmpty(repo.name) ?? this.source.repo,
      description: nullIfEmpty(repo.description),
      homepage: nullIfEmpty(repo.homepage),
      topics: stringArray(repo.topics),
      license: toSpdx(repo.license?.spdx_id),
      defaultBranch: nullIfEmpty(repo.default_branch),
      webUrl,
      cloneUrl: nullIfEmpty(repo.clone_url) ?? `${webUrl}.git`,
      issuesUrl: repo.has_issues === true ? `${webUrl}/issues` : null,
      template: repo.is_template === true,
      archived: repo.archived === true,
    };
  }

  async fetchReleases(): Promise<ImportedReleases> {
    const page = await this.client.getAll<GithubRelease>(`${this.repoUrl()}/releases?per_page=${PER_PAGE}`);
    const web = this.fallbackWebUrl();
    const releases: Release[] = [];
    for (const entry of page.items) {
      // Drafts are invisible to anonymous visitors and would appear and disappear with the
      // token used to build, which is exactly the kind of non-determinism we avoid.
      if (entry.draft === true) continue;
      const tag = nullIfEmpty(entry.tag_name);
      const publishedAt = toIsoDate(entry.published_at) ?? toIsoDate(entry.created_at);
      // Without a tag or a usable date there is nothing the site could route to or sort by.
      if (tag === null || publishedAt === null) continue;
      releases.push({
        tag,
        name: nullIfEmpty(entry.name) ?? tag,
        body: entry.body ?? '',
        url: mapUrl(nullIfEmpty(entry.html_url), web),
        prerelease: entry.prerelease === true,
        publishedAt,
        author: nullIfEmpty(entry.author?.login),
        assets: sortAssets(
          (entry.assets ?? []).map((a) => toAsset(a, web)).filter((a): a is ReleaseAsset => a !== null),
        ),
      });
    }
    return { releases: sortReleases(releases), truncated: page.truncated };
  }

  /** `<host>/repos/<owner>/<repo>`, host normalised so a trailing slash cannot double up. */
  private repoUrl(): string {
    const base = this.source.host.replace(/\/+$/, '');
    return `${base}/repos/${encodeURIComponent(this.source.owner)}/${encodeURIComponent(this.source.repo)}`;
  }

  /** Used only when the API omits `html_url`; api.github.com maps to github.com. */
  private fallbackWebUrl(): string {
    const base = this.source.host.replace(/\/+$/, '').replace(/^https:\/\/api\.github\.com$/, 'https://github.com');
    return `${base.replace(/\/api\/v3$/, '')}/${this.source.owner}/${this.source.repo}`;
  }
}

/** A provider URL made absolute against the repo page, or null when there was none. */
function mapUrl(value: string | null, base: string): string | null {
  return value === null ? null : absoluteUrl(value, base);
}

/** GitHub asset → schema asset. `browser_download_url` is the one a visitor can actually use. */
function toAsset(asset: GithubAsset, base: string): ReleaseAsset | null {
  const name = nullIfEmpty(asset.name);
  const url = nullIfEmpty(asset.browser_download_url) ?? nullIfEmpty(asset.url);
  if (name === null || url === null) return null;
  // Absolutised so a relative URL cannot render as a same-origin link on the generated site.
  return { name, url: absoluteUrl(url, base), size: byteSize(asset.size), contentType: nullIfEmpty(asset.content_type) };
}

File: internal/ingest/github.go
package ingest

import (
	"context"
	"encoding/json"
	"strings"

	"frznforge/internal/config"
	"frznforge/internal/model"
)

// Port of src/lib/importers/github.ts (REST v3). Talks to `<host>/repos/<owner>/<repo>` and
// `<host>/repos/<owner>/<repo>/releases`; host defaults to https://api.github.com and is the
// `/api/v3` base on GitHub Enterprise.

// githubPerPage is releases per request — GitHub's maximum, so most repos need a single round
// trip.
const githubPerPage = "100"

// GithubImporter reads repo metadata and releases from the GitHub REST API.
type GithubImporter struct {
	source config.RepoSourceConfig
	client *JSONClient
}

// NewGithubImporter builds a GitHub importer for one configured source.
func NewGithubImporter(source config.RepoSourceConfig, ctx ImporterContext) *GithubImporter {
	return &GithubImporter{
		source: source,
		client: NewJSONClient(JSONClientOptions{
			Auth:      AuthBearer,
			HTTP:      ctx.HTTP,
			Token:     ctx.Token,
			UserAgent: ctx.UserAgent,
			Accept:    "application/vnd.github+json",
			Backoff:   ctx.Backoff,
			Now:       ctx.Now,
		}),
	}
}

func (g *GithubImporter) Provider() string { return "github" }

// FetchMeta reads `GET /repos/{owner}/{repo}`.
func (g *GithubImporter) FetchMeta(ctx context.Context) (ImportedRepoMeta, error) {
	raw, err := g.client.Get(ctx, g.repoURL())
	if err != nil {
		return ImportedRepoMeta{}, err
	}
	repo := asObject(raw)

	webURL := g.fallbackWebURL()
	if v := NullIfEmpty(repo["html_url"]); v != nil {
		webURL = *v
	}
	cloneURL := webURL + ".git"
	if v := NullIfEmpty(repo["clone_url"]); v != nil {
		cloneURL = *v
	}
	name := g.source.Repo
	if v := NullIfEmpty(repo["name"]); v != nil {
		name = *v
	}
	var issuesURL *string
	if isTrue(repo["has_issues"]) {
		issuesURL = strPtr(webURL + "/issues")
	}
	return ImportedRepoMeta{
		Name:          &name,
		Description:   NullIfEmpty(repo["description"]),
		Homepage:      NullIfEmpty(repo["homepage"]),
		Topics:        StringArray(repo["topics"]),
		License:       ToSpdx(asObject(repo["license"])["spdx_id"]),
		DefaultBranch: NullIfEmpty(repo["default_branch"]),
		WebURL:        webURL,
		CloneURL:      cloneURL,
		IssuesURL:     issuesURL,
		Template:      isTrue(repo["is_template"]),
		Archived:      isTrue(repo["archived"]),
	}, nil
}

// FetchReleases reads `GET /repos/{owner}/{repo}/releases`, paginated.
func (g *GithubImporter) FetchReleases(ctx context.Context) (ImportedReleases, error) {
	page, err := g.client.GetAll(ctx, g.repoURL()+"/releases?per_page="+githubPerPage)
	if err != nil {
		return ImportedReleases{}, err
	}
	web := g.fallbackWebURL()
	releases := []model.Release{}
	for _, item := range page.Items {
		entry := asObject(item)
		// Drafts are invisible to anonymous visitors and would appear and disappear with the
		// token used to build, which is exactly the kind of non-determinism we avoid.
		if isTrue(entry["draft"]) {
			continue
		}
		tag := NullIfEmpty(entry["tag_name"])
		publishedAt := firstNonNil(ToISODate(entry["published_at"]), ToISODate(entry["created_at"]))
		// Without a tag or a usable date there is nothing the site could route to or sort by.
		if tag == nil || publishedAt == nil {
			continue
		}
		name := *tag
		if v := NullIfEmpty(entry["name"]); v != nil {
			name = *v
		}
		assets := []model.ReleaseAsset{}
		for _, rawAsset := range asArray(entry["assets"]) {
			if asset := githubAsset(asObject(rawAsset), web); asset != nil {
				assets = append(assets, *asset)
			}
		}
		releases = append(releases, model.Release{
			Tag:         *tag,
			Name:        name,
			Body:        rawStringOr(entry["body"], ""),
			URL:         mapURL(NullIfEmpty(entry["html_url"]), web),
			Prerelease:  isTrue(entry["prerelease"]),
			PublishedAt: *publishedAt,
			Author:      NullIfEmpty(asObject(entry["author"])["login"]),
			Assets:      SortAssets(assets),
		})
	}
	return ImportedReleases{Releases: SortReleases(releases), Truncated: page.Truncated}, nil
}

// repoURL is `<host>/repos/<owner>/<repo>`, host normalised so a trailing slash cannot double up.
func (g *GithubImporter) repoURL() string {
	base := trimTrailingSlashes(g.source.Host)
	return base + "/repos/" + encodeURIComponent(g.source.Owner) + "/" + encodeURIComponent(g.source.Repo)
}

// fallbackWebURL is used only when the API omits html_url; api.github.com maps to github.com.
func (g *GithubImporter) fallbackWebURL() string {
	base := trimTrailingSlashes(g.source.Host)
	if base == "https://api.github.com" {
		base = "https://github.com"
	}
	base = strings.TrimSuffix(base, "/api/v3")
	return base + "/" + g.source.Owner + "/" + g.source.Repo
}

// githubAsset maps a GitHub asset onto an artifact asset. browser_download_url is the one a
// visitor can actually use; download_count is deliberately ignored, because it would change the
// artifact between two builds of the same commits.
func githubAsset(asset map[string]json.RawMessage, base string) *model.ReleaseAsset {
	name := NullIfEmpty(asset["name"])
	url := firstNonNil(NullIfEmpty(asset["browser_download_url"]), NullIfEmpty(asset["url"]))
	if name == nil || url == nil {
		return nil
	}
	// Absolutised so a relative URL cannot render as a same-origin link on the generated site.
	return &model.ReleaseAsset{
		Name:        *name,
		URL:         AbsoluteURL(*url, base),
		Size:        ByteSize(asset["size"]),
		ContentType: NullIfEmpty(asset["content_type"]),
	}
}

// mapURL makes a provider URL absolute against the repo page, or keeps nil when there was none.
func mapURL(value *string, base string) *string {
	if value == nil {
		return nil
	}
	return strPtr(AbsoluteURL(*value, base))
}

// rawStringOr is `value ?? fallback` for a JSON field the schema types as a plain string.
func rawStringOr(raw json.RawMessage, fallback string) string {
	if s, ok := rawString(raw); ok {
		return s
	}
	return fallback
}

// trimTrailingSlashes is the `/\/+$/` strip every provider applies to its configured host.
func trimTrailingSlashes(s string) string { return strings.TrimRight(s, "/") }

// encodeURIComponent escapes exactly what JavaScript's encodeURIComponent escapes: everything
// outside A-Za-z0-9-_.!~*'(), byte by byte over the UTF-8 encoding, with uppercase hex.
//
// Go's url.PathEscape is not the same function — it leaves $&+,:;=@ alone — and these values
// end up in webUrl/cloneUrl, which are artifact bytes.
func encodeURIComponent(s string) string {
	const unreserved = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_.!~*'()"
	const hex = "0123456789ABCDEF"
	var b strings.Builder
	b.Grow(len(s))
	for i := 0; i < len(s); i++ {
		c := s[i]
		if strings.IndexByte(unreserved, c) >= 0 {
			b.WriteByte(c)
			continue
		}
		b.WriteByte('%')
		b.WriteByte(hex[c>>4])
		b.WriteByte(hex[c&0x0f])
	}
	return b.String()
}
